### PR TITLE
[bitv] Rewrite the canonizer

### PR DESCRIPTION
**Pull request by bclement-ocp**
This is a full rewrite of the Canonizer module in the bitvector solver. The new Canonizer (now called Canon) performs the same work as the old one, but does so slightly differently -- in particular, we now build terms lazily, which allows us to ignore parts of the tree for performance.

The main motivation is actually not performance, but the ability to integrate bitwise operators (bvnot, bvand, bvor) into the canonizer in a more natural way. This is essentially the same architecture as that of `make_aux` from #669, but lifted to use a simple custom type (`view`) to avoid building new terms during canonization and benefit from match exhaustiveness checks.

### CODE REVIEW

**Comment by bclement-ocp** [comment]
There are no regressions on the ~800 QF_BV benchmarks that we have the primitives for, and 6 benchmarks that used to timeout and now barely pass (on 4.14).

File: src/lib/reasoners/bitv.ml [hunk 1]
```diff
@@ -192,137 +192,144 @@ module Shostak(X : ALIEN) = struct
...
+      f x ctx
+    let ret x ctx = x, ctx
+    let (and*) x y =
+      let* x = x in
+      let* y = y in
+      ret (x, y)
+    let (let+) x f ctx =
+      let x, ctx = x ctx in
+      f x, ctx
+    let (and+) = (and*)
```
Comment (hra687261) [inline src/lib/reasoners/bitv.ml:228]: A bit weird to use `x` as both a function and the result of a function application, the argument `x` could be called `g` for example.

**Review by hra687261 — APPROVED** [review]
Not really what I had in mind for the canonizer but seems to be ok.

**Comment by bclement-ocp** [comment]
> Not really what I had in mind for the canonizer but seems to be ok.

Happy to settle on another approach if you have something better to suggest, what did you have in mind?

File: src/lib/reasoners/bitv.ml (hunk 1 above)
Comment (bclement-ocp) [reply src/lib/reasoners/bitv.ml:228]: I think it is relatively idiomatic to have `x` be both the "stateful" `'a m = 'st -> 'a * 'st` and the "inner" `'a`, this is what happens e.g. when we do `let* x = x in ...` as well.

I'm happy to call it `mx` here, and will add a comment explaining that `(let*)` and `(let+)` implement `bind` and `fmap` for the state monad as well.

**Comment by hra687261** [comment]
I'm referring to what was discussed in #625, although it doesn't really matter (for now at least), if that version turns out to be better I'll open a PR at some point for it.

**Comment by bclement-ocp** [comment]
Are you referring to https://github.com/OCamlPro/alt-ergo/issues/625#issuecomment-1607452057 ? If so I believe the implementation in this PR does what you described?

With `(extract 0 0 (concat complex #b0))` we never recurse into `complex` (well technically we do look at the head constructor of `complex`, but that does not depend on the size of `complex`).

**Comment by hra687261** [comment]
True, it does what is described. By "not what I had in mind", I was referring to the way in which to do it, what I had in mind was to completely remove the intermediary state (as it was said in the comment).
It's not clear that it would be better to remove the intermediary state though.

**Pull request merged**